### PR TITLE
Improve ipv6 locallink support for windows

### PR DESCRIPTION
**Pull request by pezhang**
1. Enable get host and windows guest locallink ipv6
2. Add extra notes for caculating ipv6 by mac

ID: 1566963
Signed-off-by: Pei Zhang <pezhang@redhat.com>

### CODE REVIEW

File: virttest/utils_net.py [hunk 1]
```diff
@@ -1440,12 +1437,16 @@ def get_guest_ip_addr(session, mac_addr, os_type="linux", ip_version="ipv4",
             else:
                 return None
         else:
```
Comment (sitoliu) [inline virttest/utils_net.py]: Hi Pei, what about `elif ip_version == "ipv6":`

**Comment by sitoliu** [comment]
Others looks good to me.

Acked-by: Sitong Liu <siliu@redhat.com>

File: virttest/utils_net.py [hunk 2]
```diff
@@ -1440,12 +1437,16 @@ def get_guest_ip_addr(session, mac_addr, os_type="linux", ip_version="ipv4",
...
-            global_address = [x for x in nic_address["ipv6"]
-                              if not x.lower().startswith("fe80")]
-            if global_address:
-                return global_address[0]
+            if linklocal:
```
Comment (sitoliu) [inline virttest/utils_net.py]: Pei, could you explain the reason to replace `ipv6_from_mac_addr(mac_addr)`?

File: virttest/utils_net.py [hunk 3]
```diff
@@ -3193,34 +3198,40 @@ def gen_ipv4_addr(network_num="10.0.0.0", network_prefix="24", exclude_ips=[]):
...
-def get_host_ip_address(params, ip_ver="ipv4"):
+def get_host_ip_address(params, ip_ver="ipv4", linklocal=False):
     """
     Returns ip address of host specified in host_ip_addr parameter if provided.
     Otherwise ip address on interface specified in netdst parameter is returned.
     In case of "nettype == user" "netdst" parameter is left empty, then the
     default interface of the system is used.
     :param params
     :param ip_ver: Host IP version, ipv4 or ipv6.
+    :param linklocal: Wether ip address is local or remote
```
Comment (sitoliu) [inline virttest/utils_net.py]: s/Wether/Whether

**Comment by pezhang** [comment]
Note about the ipv6 linklocal: network interface has a default ipv6 address which begins with "fe80::'. For linux os, it's using EUI-64 which calculating ipv6 by mac. For windows guest, it gets ipv6 by a random value. 

When getting ipv6, we need to specify it's linklocal or not.

@sitoliu

**Comment by pezhang** [comment]
@sitoliu @xiywang Already updated, please help to review. Thanks.

**Comment by pezhang** [comment]
@sitoliu @xiywang Already updated. Please help to review. Thanks.

**Comment by sitoliu** [comment]
@pezhang  Thanks Pei, I got it.

**Comment by xiywang** [comment]
LGTM.

Acked-by: Xiyue Wang xiywang@redhat.com

File: virttest/utils_net.py [hunk 4]
```diff
@@ -1710,6 +1713,10 @@ def create_network_script(iface_name, mac_addr, boot_proto, net_mask,
 
 def ipv6_from_mac_addr(mac_addr):
     """
+    Note:
+    Only support systems which choose EUI-64
+     - RHEL support
```
Comment (luckyh) [inline virttest/utils_net.py]: Should be `"Linux"` instead of `"RHEL"`?

File: virttest/utils_net.py [hunk 5]
```diff
@@ -1439,13 +1436,19 @@ def get_guest_ip_addr(session, mac_addr, os_type="linux", ip_version="ipv4",
...
                 return None
-        else:
-            global_address = [x for x in nic_address["ipv6"]
-                              if not x.lower().startswith("fe80")]
-            if global_address:
-                return global_address[0]
+        elif ip_version == "ipv6":
+            if linklocal:
+                return [x for x in nic_address["ipv6"]
+                        if x.lower().startswith("fe80")][0]
```
Comment (luckyh) [inline virttest/utils_net.py]: It would raise a `IndexError` if there is no link local addr.

File: virttest/utils_net.py [hunk 6]
```diff
@@ -3193,34 +3200,40 @@ def gen_ipv4_addr(network_num="10.0.0.0", network_prefix="24", exclude_ips=[]):
...
     if ip_ver == "ipv6":
         ver = netifaces.AF_INET6
     else:
         ver = netifaces.AF_INET
     addr = netifaces.ifaddresses(ifname).get(ver)
+
     if addr is not None:
-        return [a['addr'] for a in addr if not a['addr'].startswith('fe80')][0]
+        if linklocal:
+            return [a['addr'] for a in addr if a['addr'].startswith('fe80')][0]
```
Comment (luckyh) [inline virttest/utils_net.py]: Well ipv4 has link local addr in other form, so need to distinguish it with ipv6's.

see https://tools.ietf.org/html/rfc3927

**Comment by luckyh** [comment]
@balamuruhans @sathnaga request for your review, thanks.

File: virttest/utils_net.py [hunk 7]
```diff
@@ -1439,13 +1436,19 @@ def get_guest_ip_addr(session, mac_addr, os_type="linux", ip_version="ipv4",
...
+                return [x for x in nic_address["ipv6"]
+                        if x.lower().startswith("fe80")][0]
             else:
-                return None
+                global_address = [x for x in nic_address["ipv6"]
+                                  if not x.lower().startswith("fe80")]
+                if global_address:
+                    return global_address[0]
+                else:
+                    return None
```
Comment (balamuruhans) [inline virttest/utils_net.py]:
```python
try:
    return [x for x in nic_address["ipv6"]
            if not x.lower().startswith("fe80")][0]:
except IndexError:
    return None
```

Comment (balamuruhans) [reply virttest/utils_net.py]: same can be done for L1442.

File: virttest/utils_net.py [hunk 8]
```diff
@@ -3193,34 +3200,40 @@ def gen_ipv4_addr(network_num="10.0.0.0", network_prefix="24", exclude_ips=[]):
             yield str(ip_address)
 
 
-def get_ip_address_by_interface(ifname, ip_ver="ipv4"):
+def get_ip_address_by_interface(ifname, ip_ver="ipv4", linklocal=False):
     """
     returns ip address by interface
     :param ifname: interface name
     :param ip_ver: Host IP version, ipv4 or ipv6.
+    :param linklocal: Whether ip address is local or remote
```
[the remark this answers is not in the document]
Comment (pezhang) [reply virttest/utils_net.py]: @balamuruhans  You are right. Thanks for review.

File: virttest/utils_net.py (hunk 4 above)
Comment (pezhang) [reply virttest/utils_net.py]: @luckyh updated. Thanks for review.

File: virttest/utils_net.py (hunk 5 above)
Comment (pezhang) [reply virttest/utils_net.py]: @luckyh updated. Thanks.

File: virttest/utils_net.py (hunk 7 above)
Comment (pezhang) [reply virttest/utils_net.py]: @balamuruhans  Updated. Thanks.

**Comment by pezhang** [comment]
@luckyh @balamuruhans  Already updated following your suggestions. Could you please help to review again? Thanks.

File: virttest/utils_net.py [hunk 9]
```diff
@@ -1435,17 +1432,21 @@ def get_guest_ip_addr(session, mac_addr, os_type="linux", ip_version="ipv4",
...
-                return None
+            linklocal_prefix = "169.254"
+        elif ip_version == "ipv6":
+            linklocal_prefix = "fe80"
         else:
-            global_address = [x for x in nic_address["ipv6"]
-                              if not x.lower().startswith("fe80")]
-            if global_address:
-                return global_address[0]
+            raise IPAddrGetError(mac_addr, "Unknown ip version type")
```
Comment (luckyh) [inline virttest/utils_net.py]: As the exception would be caught and raised again in line 1450~1452, I'd suggest to raise `ValueError` at here.

**Comment by pezhang** [comment]
@luckyh Updated. Please help to review again. Thanks.

File: virttest/utils_net.py [hunk 10]
```diff
@@ -3193,34 +3198,55 @@ def gen_ipv4_addr(network_num="10.0.0.0", network_prefix="24", exclude_ips=[]):
...
+                                "%s !", ifname)
+                return None
+        else:
+            try:
+                return [a['addr'] for a in addr
+                        if not a['addr'].lower().startswith(linklocal_prefix)][0]
+            except IndexError:
+                logging.warning("No IP address configured for "
+                                "the network interface %s !", ifname)
+                return None
```
Comment (balamuruhans) [inline virttest/utils_net.py]: you can handle 3218-3234 in the same way as you have handled in 1441-1449. it helps to reduce no of lines.

**Pull request merged**